sled-agent/src/services.rs

```diff
@@ -2878,211 +2927,224 @@ impl ServiceManager {
                 // Time is synchronized
                 Ok(TimeSync { sync: true, .. }) => true,
                 // Time is not synchronized, or we can't check
                 _ => false,
             };
-        for zone in zones_to_be_added.clone() {
-            if zone_requires_timesync(&zone.zone_type) && !time_is_synchronized
-            {
-                return Err(Error::TimeNotSynchronized);
-            }
+
+        // Concurrently boot all new zones
+        let StartZonesResult { zones, errors } = self
+            .start_omicron_zones(
+                zones_to_be_added,
+                time_is_synchronized,
+                &all_u2_pools,
+            )
+            .await?;
+
+        // Add the new zones to our tracked zone set
+        existing_zones.extend(
+            zones.into_iter().map(|zone| (zone.name().to_string(), zone)),
+        );
+
+        // If any zones failed to start, exit with an error
+        if !errors.is_empty() {
+            return Err(Error::ZoneEnsure { errors });
         }
+        Ok(())
+    }
 
-        // Destroy zones that should not be running
-        for zone in zones_to_be_removed {
-            let expected_zone_name = zone.zone_name();
-            if let Some(mut zone) = existing_zones.remove(&expected_zone_name) {
-                debug!(
+    // Attempts to take a zone bundle and remove a zone.
+    //
+    // Logs, but does not return an error on failure.
+    async fn zone_bundle_and_try_remove(
+        &self,
+        existing_zones: &mut MutexGuard<'_, ZoneMap>,
+        zone: &OmicronZoneConfig,
+    ) {
+        let log = &self.inner.log;
+        let expected_zone_name = zone.zone_name();
+        if let Some(mut zone) = existing_zones.remove(&expected_zone_name) {
+            debug!(
+                log,
+                "removing an existing zone";
+                "zone_name" => &expected_zone_name,
+            );
+            if let Err(e) = self
+                .inner
+                .zone_bundler
+                .create(&zone.runtime, ZoneBundleCause::UnexpectedZone)
+                .await
+            {
+                error!(
                     log,
-                    "removing an existing zone";
+                    "Failed to take bundle of unexpected zone";
                     "zone_name" => &expected_zone_name,
+                    "reason" => ?e,
                 );
-                if let Err(e) = self
-                    .inner
-                    .zone_bundler
-                    .create(&zone, ZoneBundleCause::UnexpectedZone)
-                    .await
+            }
+            if let Err(e) = zone.runtime.stop().await {
+                error!(log, "Failed to stop zone {}: {e}", zone.name());
+            }
+        } else {
+            warn!(log, "Expected to remove zone, but could not find it");
+        }
+    }
+
+    // Ensures that if a zone is about to be installed, it does not exist.
+    async fn ensure_removed(
```

Comment:
This method is *almost* the same as it was before. We used to only check `existing_zones` to decide which zones to look at, and then we used to "ignore" running zones, because if it's already up, the zone must have already been successfully launched by someone.

In retrospect, I think this wasn't the right choice. We only called this code in a context where we want to **explicitly** add a new zone that we don't know about (see: `zones_to_be_added`), and the goal is to remove detritus from the system ("Get to a clean slate") before trying to start a new zone. Trying to handle the state of "the zone could be okay, or it could be bad an need removal, and we need to decide which is which" made this hard to reason about.

In reality, the calls to `zone_bundle_and_try_remove` for `zones_to_be_removed` should be the "normal" way of removing zones, but this slightly updated method just acts as a conservative back-stop to clear anything that could prevent a new zone from being started.